### PR TITLE
Introduce test_register_custom_op

### PR DESCRIPTION
**Pull request by jinevening**
This adds a test file for register_custom_op.py and fixes some bugs.

TICO-DCO-1.0-Signed-off-by: Hyukjin Jeong <hj1.jeong@samsung.com>

---
Related to: https://github.com/Samsung/TICO/issues/322

### CODE REVIEW

File: tico/utils/register_custom_op.py [hunk 1]
```diff
@@ -31,9 +31,11 @@ def resize_nearest_neighbor(input_: torch.Tensor, size: List[int]) -> torch.Tens
         W_scale_factor = size[2] / W
         if H_scale_factor != W_scale_factor:
             raise RuntimeError("Scale factor of H and W should be same.")
-        return torch.nn.functional.interpolate(
-            input_, scale_factor=H_scale_factor, mode="nearest"
+        permuted = torch.permute(input_, [0, 3, 1, 2])
+        resized = torch.nn.functional.interpolate(
+            permuted, scale_factor=H_scale_factor, mode="nearest"
         )
+        return torch.permute(resized, [0, 2, 3, 1])
```
Comment (jinevening) [inline tico/utils/register_custom_op.py:38]: Existing kernel has a bug that does not consider layout conversion.

**Comment by jinevening** [comment]
Test coverage is increased from 59% -> 74%

File: test/unit_test/utils_test/test_register_custom_op.py [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+        with self.assertRaises(RuntimeError) as context:
+            torch.ops.circle_custom.resize_nearest_neighbor(input_tensor, size)
+
+        self.assertIn("Scale factor of H and W should be same", str(context.exception))
+
+    def test_circle_conv2d_with_defaults(self):
+        """Test CircleConv2d with default parameters"""
+        # This should cover lines 74-94 in register_custom_op.py
+        input_tensor = torch.randn(1, 32, 32, 3)  # NHWC
+        weight_tensor = torch.randn(16, 3, 3, 3)  # OHWI
```
Comment (miusic) [inline test/unit_test/utils_test/test_register_custom_op.py:66]: May I ask why you didn't use `self.input_tensor` and `self.weight_tensor`?

File: test/unit_test/utils_test/test_register_custom_op.py [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+            input_tensor, weight_tensor, padding="same"
+        )
+
+        # Check output shape with "same" padding
+        expected_shape = [1, 32, 32, 16]  # NHWC with "same" padding
+        self.assertEqual(list(result.shape), expected_shape)
+
+    def test_circle_depthwise_conv2d_basic(self):
+        """Test CircleDepthwiseConv2d basic functionality"""
+        # This should cover lines 203-248 in register_custom_op.py
```
Comment (miusic) [inline test/unit_test/utils_test/test_register_custom_op.py:127]:
```suggestion
```
IMHO, this comment looks redundant and easy to break when `register_custom_op.py` is modified

File: test/unit_test/utils_test/test_register_custom_op.py (hunk 2 above)
Comment (jinevening) [reply test/unit_test/utils_test/test_register_custom_op.py:61]: Thanks for finding. Those attributes are unused. I've removed them.

File: test/unit_test/utils_test/test_register_custom_op.py [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+        # Test with different batch sizes
+        for batch_size in [1, 2, 4]:
+            input_tensor = torch.randn(batch_size, 16, 16, 3)
+            weight_tensor = torch.randn(8, 3, 3, 3)
+
+            result = torch.ops.circle_custom.conv2d(input_tensor, weight_tensor)
+            expected_batch_size = batch_size
+            self.assertEqual(result.shape[0], expected_batch_size)
+
+    def test_custom_ops_with_different_data_types(self):
```
Comment (miusic) [inline test/unit_test/utils_test/test_register_custom_op.py:387]: If just the `circle_custom.conv2d` is used for this test, IMHO, it would be more proper to rename it as a test for conv2d, like `test_circle_conv2d_other_dtype`

File: test/unit_test/utils_test/test_register_custom_op.py [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+        padding = [1, 1]
+        dilation = [1, 1]
+        ceil_mode = True
+
+        result = torch.ops.circle_custom.maxpool2d(
+            input_tensor, kernel_size, stride, padding, dilation, ceil_mode
+        )
+
+        # Check output shape with custom params
+        expected_shape = [1, 32, 32, 3]  # NHWC with kernel_size=3, stride=1, padding=1
```
Comment (miusic) [inline test/unit_test/utils_test/test_register_custom_op.py:244]:
```suggestion
        expected_shape = [1, 32, 32, 3]  # NHWC with kernel_size=3, stride=1, padding=1, dilation=1
```
(optional)
It is not that critical, but there is no description for `dilation`

Honestly, I think that these kinds of comments are not so informative, so it has no problem to be removed 🤔
```suggestion
        expected_shape = [1, 32, 32, 3]
```

File: test/unit_test/utils_test/test_register_custom_op.py [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+        """Test custom ops with conv2d different data types"""
+        # Test with different data types
+        for dtype in [torch.float32, torch.float16]:
+            input_tensor = torch.randn(1, 16, 16, 3, dtype=dtype)
+            weight_tensor = torch.randn(8, 3, 3, 3, dtype=dtype)
+
+            result = torch.ops.circle_custom.conv2d(input_tensor, weight_tensor)
+            self.assertEqual(result.dtype, dtype)
+
+    def test_custom_ops_edge_cases(self):
```
Comment (miusic) [inline test/unit_test/utils_test/test_register_custom_op.py:397]: I'm not sure whether it is a true edge case or not.. 🤔
The tensor with all zero values is not a regular case, but the zero value itself is not a special, I guess
